src/Source/Framework/Bio.Core/IO/BAM/BAMParser.cs

```diff
@@ -954,7 +971,8 @@ private SAMAlignedSequence GetAlignedSequence(int start, int end)
             int blockLen = Helper.GetInt32(array, 0);
             byte[] alignmentBlock = new byte[blockLen];
             ReadUnCompressedData(alignmentBlock, 0, blockLen);
-            SAMAlignedSequence alignedSeq = new SAMAlignedSequence();
+			T alignedSeq = GetNewAlignedSequence();
```

Comment:
Good question. Using "new T()" means that every time this line of code gets executed, a new object will be created. However, by abstracting the object creation, we have created an inversion of control for the creation of objects. The protected virtual GetNewAlignedSequence() method provides an easy hook for allowing objects of type T to come from a factory pattern, or even a pool-type pattern for object creation. 

Returning pooled SAMAlignedSequence objects from a BAM file is important because such files generally contain millions of reads, and the content of such files often do not fit in memory. Furthermore, instead of creating a new object for each read, then relying on garbage collection to de allocate memory, it is often better to take control of memory usage and object creation in our own managed code. 

A simple, but powerful, use case that I have in mind for this change is to create a derived class where GetNewAlignedSequence always returns the same object instance. This class would be called something like BamStreamingParser, and it will be a tool that I can use to enumerate reads in a BAM file with an O(1) memory footprint. A slightly more sophisticate (and thread-safe) version of this class would return a small number of object instances from a pool and would ultimately have a memory footprint of O(t) where t is the number of threads pulling reads from the BAM file.